### PR TITLE
FI-2342 Token introspection

### PR DESCRIPTION
**Pull request by arscan**
# Summary
Adds Token Introspection tests to the SMART App Launch Test Kit.  It adds an addition test group, which provides some flexibility in testing and allows users to choose which specific tests to run.  It works ok, but does require the user to read up on how to use the tests before running them.

One notable aspect of this is that the 'full test' option includes generating another token for introspection, instead of just asking the user to run one of the existing sequences to get the token.  We had to do this instead of reusing one of the existing sequences is that token refresh tends to 'mess up' things a bit for these tests because we need to know what the original token response was, not the token response for the token refresh step (to get info about the id_token, I believe).  Its not a huge deal, but it makes me wonder about how all these should be organized in general.

FYI @alisawallace

### CODE REVIEW

File: lib/smart_app_launch/smart_stu2_suite.rb [hunk 1]
```diff
@@ -204,5 +221,66 @@ class SMARTSTU2Suite < Inferno::TestSuite
               }
             }
     end
+  
+    group do
```
Comment (Jammjammjamm) [inline lib/smart_app_launch/smart_stu2_suite.rb:225]: This group should live in its own file.

File: lib/smart_app_launch/token_introspection_access_token_group.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+require_relative 'standalone_launch_group_stu2'
+
+module SMARTAppLaunch
+  class TokenIntrospectionAccessTokenGroup < Inferno::TestGroup
+    title 'Request New Access Token to Introspect'
+    run_as_group
+
+    id :token_introspection_access_token_group
```
Comment (Jammjammjamm) [inline lib/smart_app_launch/token_introspection_access_token_group.rb:8]: I'd put `smart` at the start of all the ids.

File: lib/smart_app_launch/app_redirect_test_stu2.rb [hunk 3]
```diff
@@ -16,7 +16,7 @@ class AppRedirectTestSTU2 < AppRedirectTest
     )
 
     input :authorization_method,
-          title: 'Authorization Method',
+          title: 'Authorization Request Method',
```
[the remark this answers is not in the document]
Comment (arscan) [reply lib/smart_app_launch/app_redirect_test_stu2.rb:19]: Ok, I'm leaving it as-is unless @Jammjammjamm would prefer we change it back.

File: lib/smart_app_launch/smart_stu2_suite.rb (hunk 1 above)
Comment (arscan) [reply lib/smart_app_launch/smart_stu2_suite.rb]: Moved

File: lib/smart_app_launch/token_introspection_access_token_group.rb (hunk 2 above)
Comment (arscan) [reply lib/smart_app_launch/token_introspection_access_token_group.rb]: renamed

File: lib/smart_app_launch/app_redirect_test_stu2.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jammjammjamm) [reply lib/smart_app_launch/app_redirect_test_stu2.rb:19]: I don't see an issue with this.

File: lib/smart_app_launch/token_introspection_group.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      can support it, as outputs from one group will feed the inputs of the next group. However, test groups can be run
+      independently if needed.
+
+      See the individual test groups for more details and guidance.
+    )
+    group from: :smart_token_introspection_access_token_group
+    group from: :smart_token_introspection_request_group
+    group from: :smart_token_introspection_response_group
+
+    input_order :well_known_introspection_url, :custom_authorization_header, :optional_introspection_request_params,
```
Comment (Jammjammjamm) [inline lib/smart_app_launch/token_introspection_group.rb:47]: I don't know if it's intentional, but this doesn't include all of the inputs for this group.

**Review by Jammjammjamm — APPROVED** [review]
Screenshot 2023-12-08 at 12 43 29 PM

The auth and token url inputs don't have titles or descriptions. They don't show up in the launch groups because they're set by the discovery tests in those groups.

**Pull request merged**